_posts/2018-04-15-lazy-loading-images.md

```diff
@@ -0,0 +1,205 @@
+        alt="book collection photographed from top">
+    </li>
+    <li class="image__list-item">
+      <img class="image__list-image image__list-lazy-image" data-src="http://res.cloudinary.com/dsteinel/image/upload/c_scale,w_450/v1523796300/pexels-photo-559768.jpg"
+        alt="brown meadow with a tower in the very back">
+    </li>
+  </ul>
+</div>
+{% endhighlight %}
+
+{% highlight css %}
+.image-container {
+  margin: 0 auto;
+  width: 450px;
+}
+
+.image__list-lazy-image { 
+  display: none;
+}
+
+.image__list-image {
+  display: block;
+}
+
+.image__list-item {
+  margin-bottom: 40px;
+}
+{% endhighlight %}
+
+Let’s get to JavaScript part. There is a very nice little helper which was announced on the 14 September’17 by Google. They describe it as follows:
+"This specification describes an API that can be used to understand the visibility and position of DOM elements ("targets") relative to a containing element or to the top-level viewport ("root"). The position is delivered asynchronously and is useful for understanding the visibility of elements and implementing pre-loading and deferred loading of DOM content.” [source](https://www.w3.org/TR/intersection-observer/)
+
+**This is just perfect for lazy loading images!**
+
+So we first have to gather all the images we marked above as “lazy-images”. Those are the images which have a data attribute which carries our image source. Make sure that the image has no src attribute set, otherwise, the browser will load an image before the IntersectionObserver can do its magic. The logic result we are now trying to achieve is that no lazy load image has an image source, meaning that no image will be loaded initially. The imageObserver will now check on each image, if it is in the viewport. If it is, then we will replace the content of the src attribute with the content of the data-src attribute (the actual image source). If this is done, we will download and display the image.
+
+{% highlight JavaScript %}
+const options = {
+  rootMargin: '0px 0px 250px 0px'
+}
+
+// takes a callback as first parameter and options as second
+const observer = new IntersectionObserver(lazyLoader, options);
+
+// select all our image which we want to have lazy loaded
+document.querySelectorAll('.image__list-lazy-image').forEach(image => {
+  // put them in the observer
+  observer.observe(image);
+});
+
+function lazyLoader (entries) {
+  // loop through all images
+  entries.forEach(entry => {
+    // does the viewport hit the current image
+    if (entry.isIntersecting) {
+      // yes! load the image
+      lazyLoadImage(entry);
+    }
+  });
+}
+
+function lazyLoadImage (observedImage) {
+  /**
+  *  We hit an observed image!
+  *  First, remove the lazy loading class. This will show the image
+  *  Then remove the observer
+  */
+  observedImage.target.classList.remove('image__list-lazy-image');
+  
+  // does it have a data attribute with image-src?
+  if (observedImage.target.dataset.src) {
```

Comment:
`dataset` is supported in IE11+ only, you might want to add a comment to polyfill it

Reply:
or use `.getAttribute('data-src')` instead

Reply:
wow that's a good catch! Did not know that 👍